website/blog/2023-10-21-incremental-decoding/index.md

```diff
@@ -0,0 +1,71 @@
+---
+authors: [ gyxlucy ]
+
+tags: [tech design]
+
+---
+# Incremental Decoding for Good
+
+In the context of the Transformer model, which is widely used across LLMs, ***decoding*** refers to the process of generating an output sequence from an encoded input. Tabby recenty  [implemented ***incremental decoding***](https://github.com/TabbyML/tabby/pull/491) as the default decoding method. This blog will provide an introduction into why Tabby chooses incremental decoding for code completion tasks 🛠️💡.
+
+
+## Common Decoding Methods
+
+Here's an example to facilitate understanding different decoding methods:
+
+Let's say a developer starts to write a list comprehension in Python to get even numbers from a list:
+
+```python
+numbers = [1, 2, 3, 4, 5]
+evens = [x for x in numbers
+```
+
+To simplify the scenario, we assume that the language model maintains a probaility distribution as shown below,
+
+![probability](./probability.png)
+
+Here's how different decoding methods might suggest 🔍:
+
+
+### Greedy Decoding 🏆
+
+Greedy decoding selects the most probable next token at each step, which is most intuitive method but can sometimes lead to sub-optimal sequences. This is because it only considers one token at a time and makes such choice greedily. 
+
+In this particular case, greedy decoding would complete the code with `]\n print` as each of the token here has maximum probability given the chosen token before.
+
+![greedy](./greedy.png)
+
+### Beam Search 🌈
+Beam search maintains multiple possible sequences (beams) and expands them, which offers better quality at the expense of higher computation.
+
+Assuming `num_beams=2`, in this case, it'll actually produce ` if x % 2`, as `0.4*0.7=0.28` gives the highest probability when considering a sequence of 2.
+
+![beam](./beam.png)
```

Comment:
```suggestion

### Beam Search 🌈
Beam search maintains multiple possible sequences (beams) and expands them, which offers better quality at the expense of higher computation.

Assuming `num_beams=2`, in this case, it'll actually produce ` if x % 2`, as `0.4*0.7=0.28` gives the highest probability when considering a sequence of 2.

![beam](./beam.png)

### Greedy Decoding 🏆

Greedy decoding selects the most probable next token at each step, which is most intuitive method but can sometimes lead to sub-optimal sequences. This is because it only considers one token at a time and makes such choice greedily. 

In this particular case, greedy decoding would complete the code with `]\n print` as each of the token here has maximum probability given the chosen token before.

![greedy](./greedy.png)
```